### PR TITLE
Feature/improve notification infrastructure

### LINKED ISSUE #508: Enhance the notifications infrastructure
As a developer, I want to have advanced infrastructure including BloC layer so that I can handle the push notifications more easily.

https://pub.dev/packages/rx_bloc_cli/example#push-notifications

- [x] Create push notifications bloc
- [x] Create an example of user navigation on notification tapped
- [x] Display user notification settings in a switch widget (if the user enabled/disabled push notifications)

### PR DESCRIPTION
**Pull request by JvnSlv**
Close #508

### CODE REVIEW

**Review by DDavidPrime — COMMENTED** [review]
While testing I noticed two issues:
1. Even though the notifications were initially disabled, I was still receiving them.
2. When I tried turning the notifications on in the profile page, I kept receiving an error and I was unable to turn it on (which in result meant that I wasn't able to receive any notifications).

Video attachment:
https://drive.google.com/file/d/1fYMci6qq9LmtHkF-gMFtwWKXVVZFQQuq/view?usp=share_link

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+@RxBloc()
+class PushNotificationsBloc extends $PushNotificationsBloc {
+  PushNotificationsBloc(this._routerBloc) {
+    onRouting.connect().addTo(_compositeSubscription);
+  }
+
+  final RouterBlocType _routerBloc;
+
+  @override
+  ConnectableStream<void> _mapToOnRoutingState() => Rx.merge([
```
Comment (PrimeSwiftArt) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart:40]: Rx.merge is not needed if there is only one stream for merging.

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/// A contract class containing all states of the PushNotificationsBloC.
+abstract class PushNotificationsBlocStates {
+  ConnectableStream<void> get onRouting;
+
+  /// The loading state
+  Stream<bool> get isLoading;
+
+  /// The error state
+  Stream<ErrorModel> get errors;
```
Comment (PrimeSwiftArt) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart:28]: This states don't emit anything because this bloc is not set as handler anywhere.

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+/// A contract class containing all events of the PushNotificationsBloC.
+abstract class PushNotificationsBlocEvents {
+  /// Event for handling opening push notifications from background
+  /// or from in-app
+  void tapOnEvent(NotificationModel event);
+}
+
+/// A contract class containing all states of the PushNotificationsBloC.
+abstract class PushNotificationsBlocStates {
+  ConnectableStream<void> get onRouting;
```
Comment (PrimeSwiftArt) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart:22]: What is the purpose of this state? It is not consumed anywhere.

Comment (JvnSlv) [reply packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart:21]: This state is used for handling routing on push notification received

Comment (PrimeSwiftArt) [reply packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart:21]: But is not listened anywhere. In this case it must be a private instance property of Bloc class.

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import 'package:rxdart/rxdart.dart';
+
+import '../../lib_router/blocs/router_bloc.dart';
+import '../../lib_router/router.dart';
+import '../models/notification_model.dart';
+part 'push_notifications_bloc.rxb.g.dart';
+
+/// A contract class containing all events of the PushNotificationsBloC.
+abstract class PushNotificationsBlocEvents {
+  void tapOnEvent(NotificationModel event);
```
Comment (PrimeSwiftArt) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart:13]: Why did you remove this documentation comment?
```
  /// Event for handling opening push notifications from background
  /// or from in-app
```

Comment (JvnSlv) [reply packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_blocs/push_notifications_bloc.dart:15]: Removed by mistake 👍

**Review by DDavidPrime — APPROVED** [review]

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/pubspec.yaml [hunk 5]
```diff
@@ -32,6 +32,7 @@ dependencies:
   jaguar_jwt: ^3.0.0 # This package is used for the local server (if not used, can be removed)
   json_annotation: ^4.7.0
   jwt_decoder: ^2.0.1
+  notification_permissions: ^0.6.1
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/pubspec.yaml:35]: I'm wondering won't it be better to use: https://pub.dev/packages/permission_handler for the permissions? It appears to require some initial setup, but is well supported and can also be used in the future if other permissions need to be included in the app.

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/data_sources/local/profile_local_data_source.dart [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+/// This class is used to save user profile related settings into shared preferences
+/// Currently only notification settings are saved
+class ProfileLocalDataSource {
+  static const _notificationsEnabled = 'notificationsEnabled';
+  static const _notificationsSubscribed = 'notificationsSubscribed';
+  ProfileLocalDataSource(this._sharedPreferences);
+
+  final SharedPreferencesInstance _sharedPreferences;
+
+  Future<bool> notificationsEnabled() async =>
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/data_sources/local/profile_local_data_source.dart:14]: I think currently the `notificationsEnabled()` is not being used anywhere, so it can be removed (along with `setNotificationsEnabled`.

File: packages/rx_bloc_cli/mason_templates/bricks/lib_auth/__brick__/lib/lib_auth/services/user_account_service.dart [hunk 7]
```diff
@@ -62,7 +62,10 @@ class UserAccountService {
     try {
       final pushToken =
           await _pushSubscriptionRepository.getToken(vapidKey: webVapidKey);
-      if (pushToken != null) {
+      final notificationsSubscribed =
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/lib_auth/__brick__/lib/lib_auth/services/user_account_service.dart:65]: Is there a reason why we need to keep the `notificationsSubscribed` flag? In my opinion the **pushToken** should always be sent to the backend if available. In cases where the user doesn't want to receive notifications, they can be disabled from the device's settings menu.

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/bin/server/controllers/push_notifications_controller.dart [hunk 8]
```diff
@@ -80,9 +82,12 @@ class PushNotificationsController extends ApiController {
       message,
       BadRequestException('Push message can not be empty.'),
     );
+    if (!(_pushTokens.tokens.any((element) => element.token == pushToken))) {
+      throw BadRequestException('Notifications disabled by the user');
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/bin/server/controllers/push_notifications_controller.dart:86]: This is pretty minor - usually `400 Bad Request` is used to indicate a client error like malformed request syntax. I think something like `404 Not Found` or `410 Gone` will be more suitable, since the push token is missing.

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/data_sources/local/profile_local_data_source.dart (hunk 6 above)
Comment (JvnSlv) [reply packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/data_sources/local/profile_local_data_source.dart]: Yes, you are right, I have removed it

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/pubspec.yaml (hunk 5 above)
Comment (JvnSlv) [reply packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/pubspec.yaml]: Swapped notification_permissions with permission_handler 👍

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/ios/Podfile [hunk 9]
```diff
@@ -41,8 +41,12 @@ end
 post_install do |installer|
   installer.pods_project.targets.each do |target|
     flutter_additional_ios_build_settings(target)
-    target.build_configurations.each do |config|
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/ios/Podfile:44]: This line should not be deleted - I assume you removed it by accident 😄

Comment (JvnSlv) [reply packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/ios/Podfile:44]: Yes 😅

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_services/push_notifications_service.dart [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    await _pushNotificationRepository.unsubscribe(token);
+    await _pushNotificationRepository
+        .setNotificationSubscribed(setNotifications);
+  }
+
+  Future<bool> areNotificationsEnabled() async {
+    final permissionStatus = await Permission.notification.status;
+    final shouldSubscribe =
+        await _pushNotificationRepository.notificationsSubscribed();
+    if (permissionStatus == PermissionStatus.granted && shouldSubscribe) {
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_services/push_notifications_service.dart:30]: I think for enums like `PermissionStatus` it's a good practice to use `switch` and list all the cases explicitly. This way if there's any change to the statuses in the future, we'll get compile-time feedback.

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_services/push_notifications_service.dart [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    if (token == null) {
+      throw ErrorModel();
+    }
+    await _pushNotificationRepository.unsubscribe(token);
+    await _pushNotificationRepository
+        .setNotificationSubscribed(setNotifications);
+  }
+
+  Future<bool> areNotificationsEnabled() async {
+    final permissionStatus = await firebaseMessaging.getNotificationSettings();
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_services/push_notifications_service.dart:29]: Since all the interactions with firebase messaging happen through the `_pushNotificationRepository` maybe we can add the `getNotificationSettings()` method there to be consistent?

File: packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_services/push_notifications_service.dart [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    final shouldSubscribe =
+        await _pushNotificationRepository.notificationsSubscribed();
+    switch (permissionStatus.authorizationStatus) {
+      case (AuthorizationStatus.authorized):
+        if (shouldSubscribe) {
+          await subscribe();
```
Comment (PetkovPrime) [inline packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/common_services/push_notifications_service.dart:35]: Also I'm not quite sure if this method is the best place to perform the `subscribe` and `unsubscribe` operations. The name `areNotificationsEnabled()` suggests that we are only reading the current state, without performing any underlying modifications.

**Review by PrimeSwiftArt — APPROVED** [review]

**Review by PetkovPrime — APPROVED** [review]
